### PR TITLE
Umbraco Dictionary Value Converter

### PR DESCRIPTION
**Pull request by leekelleher**
Revisiting #18, #22 (and various comments in #36)

Combining a custom attribute with a custom converter, we can populate a model's property with a value from an Umbraco dictionary-item.

For example...

```csharp
[UmbracoDictionaryValue("MyLabel")]
[TypeConverter(typeof(DittoDictionaryValueConverter))]
public string SomeLabel { get; set; }
```

There are a couple of implementation decisions that I'd like to discuss with this PR...

* The `DittoConverterWithAttribute` base/abstract converter.
  * I did it this way, as I thought we could reuse it when we do a similar attribute/converter for Relations, (see #36)
  * I'm not 100% happy with the naming. It should have "Converter" suffix, would "DittoWithAttributeConverter" make sense? :persevere:

* I've ended up using `UmbracoHelper` within `DittoDictionaryValueConverter`, as the Umbraco core's `CultureDictionaryFactoryResolver` is marked as internal in v6.2.5 :weary: (I considered using `uQuery` static helper, but that's probably worse these days)

Open for comment and suggestions. Once we're happy with the approach, we can do similar for Relations.

### CODE REVIEW

**Comment by mattbrailsford** [comment]
1) DittoAttributeDependantConverter?
2) Sounds fair to use UmbracoHelper when no other option. As long as it's self contained within the type converter it shouldn't be a problem.

One thing I am wondering is if there is anyway not to use 2 attributes. Now, that's not to say never use the UmbracoDictionaryValueAttribute, but given you are using DittoDictionaryValueConverter, shouldn't it's first action be to look up a dictionary item with a key of SomeLabel but if the UmbracoDictionaryValueAttribute is present, then use the key specified there? Not sure how this would affect the Relations datatype though.

**Comment by leekelleher** [comment]
Thanks @mattbrailsford.

1. :+1: (I like it ... still open to suggestions)
2. Yeah, I'd prefer not to use it, but there aren't many alternatives for getting dictionary item values.

re: Using one attribute - I originally did this in PR #22. But it meant adding another specific check in the `.As()` extension method, which was starting to get a bit bloated.  We discuss it a bit more #36.

**Comment by mattbrailsford** [comment]
Ok, I see what you mean. I'm just wondering though, could we have either a base class or an interface for "AlternativeValueProvider" attributes that have a single method "public object GetValue()" then grab the attribute based on the base class, thus we could have many alternative value provider attributes, but they all get handled in one claws?

**Comment by JimBobSquarePants** [comment]
:+1: On falling back to a default label.

Relations are going to have a single string key also as far as I can figure from looking at some of @Hendy's code for a relations converter. Perhaps checking if such a key exists in both implementations overwriting `CanConvertFrom` and falling back to checking if an attribute exists in the base.

I dunno how expensive those calls are though.

**Comment by mattbrailsford** [comment]
There you go, there is an overload for GetCustomAttribute that can search the ancestor hierarchy https://msdn.microsoft.com/en-us/library/4x2fs7yd%28v=vs.110%29.aspx

**Comment by JimBobSquarePants** [comment]
I use the equivalent overload for the GetCustomAttributes here. https://github.com/leekelleher/umbraco-ditto/blob/develop/src/Our.Umbraco.Ditto/Extensions/PublishedContentExtensions.cs#L391

**Comment by mattbrailsford** [comment]
My last comment was wrong, the boolean flag checks for attributes on ancester types which we don't want, but look at the docs for the standard GetCustomAttribute, it looks like the type check checks the base types anyways

https://msdn.microsoft.com/en-us/library/9zhcty09(v=vs.110).aspx

attributeType
Type: System.Type
The type, or a base type, of the custom attribute to search for.

File: src/Our.Umbraco.Ditto/ComponentModel/TypeConverters/DittoConverterWithAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+            return base.CanConvertFrom(context, sourceType);
+        }
+
+        /// <summary>
+        /// Gets the attribute.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <returns></returns>
+        public T GetAttribute(ITypeDescriptorContext context)
```
Comment (JimBobSquarePants) [inline src/Our.Umbraco.Ditto/ComponentModel/TypeConverters/DittoConverterWithAttribute.cs]: protected virtual :wink:

Comment (leekelleher) [inline src/Our.Umbraco.Ditto/ComponentModel/TypeConverters/DittoConverterWithAttribute.cs]: Doh, of course!

**Comment by leekelleher** [comment]
@mattbrailsford I think I know what you mean, but it's not quite clicked yet. Are you suggesting that we have our own Ditto version of `TypeConverterAttribute`, so then we can inherit from and extend?

One problem is that `TypeConverterAttribute` is marked as sealed :rage4:

**Comment by mattbrailsford** [comment]
No, I'm suggesting we have the UmbracoDictionaryValue attribute extend an
abstract class like AlternateValueAttribute which has a method GetValue
that must be implemented by the concrete class (in this case it would look
it up in the dictionary). We then add logic in .As() to see if any
attributes are defined that inherit AlternatValueAttribute and if so, call
the GetValue method and then use that as the property value.

What I'm suggesting then is we use the attributes to define alternative
value sources, and type converters to convert value types into something
more meaningful (if you think about it, UmbracoProperty could inherit the
base class too as its resolving a different value). We can do similar to
type converters and pass in a context.

I personally think this makes it real clean and separates things nicely.

Attribute = Resolve alternate value
TypeConverter = convert value to something else

Matt

**Comment by leekelleher** [comment]
@mattbrailsford :tada: **click** :smile: Gotcha!  I likes it!

**Comment by mattbrailsford** [comment]
ValueResolverAttribute ?

abstract object ResolveValue(DittoResolverContext context)

:)

**Comment by leekelleher** [comment]
Since we'd want to prefix it with "Ditto", how about simplifying it to "DittoValueAttribute"?
With an abstract "GetValue" method? (with whatever context-y things we want to pass in)

**Comment by leekelleher** [comment]
Closing PR.

The discussion has moved over to PR #65 (on-going at the time of writing), the `DittoValueAttribute` feels a better way forwards.

**Comment by mattbrailsford** [comment]
IMO it's fine because it's limited to a single purpose, and it's all self contained. It is responsible for resolving an Umbraco property, no more, no less.

**Pull request closed without merging**